packages/pwa-buildpack/lib/BuildBus/__tests__/BuildBus.spec.js

```diff
@@ -4,171 +4,259 @@ const pertain = require('pertain');
+            const ordered = Math.max(targets.own.takeOrders.call(0), 0);
+            const bottlesLeft = Math.max(bottles - ordered);
+            sing(
+                `${ordered} down, pass em around\n${bottlesLeft} bottles of beer on the wall!\n`
+            );
+            if (bottlesLeft) {
+                targets.own.beersLeft.call(bottlesLeft);
+            }
         });
-    })
-);
+    }
+});
+
+const allDeps = [...dependencies.entries()];
 
-pertain.mockImplementation((_, subject) => {
+// virtual module mocker
+const phasePath = (depName, phase) => `/path/to/${depName}/${phase}`;
+const doMockTargetDep = (depName, phases) =>
+    Object.entries(phases).forEach(([phase, fn]) =>
+        jest.doMock(phasePath(depName, phase), () => fn, {
+            virtual: true
+        })
+    );
+const doMockAdditionalDep = (depName, phases) => {
+    const allPhases = Object.assign({ declare() {}, intercept() {} }, phases);
+    if (process.env.BUILDBUS_DEPS_ADDITIONAL) {
+        process.env.BUILDBUS_DEPS_ADDITIONAL += `,${depName}`;
+    } else {
+        process.env.BUILDBUS_DEPS_ADDITIONAL = depName;
+    }
+    doMockTargetDep(depName, allPhases);
+};
+for (const [depName, phases] of allDeps) {
+    doMockTargetDep(depName, phases);
+}
+pertain.mockImplementation((context, subject, getDeps) => {
     const phase = subject.split('.').pop();
-    const deps = mockHandlerEntries.filter(([, subjects]) => subjects[phase]);
-    return deps.map(([depName]) => ({
+    const deps = getDeps(
+        allDeps
+            .filter(([, subjects]) => subjects[phase])
+            .map(([depName]) => depName),
+        context,
+        subject
+    );
+    return deps.map(depName => ({
         name: depName,
-        path: `/path/to/${depName}/${phase}`,
+        path: phasePath(depName, phase),
         subject
     }));
 });
 
-beforeEach(() => {
-    BuildBus.clearAll();
-    handlerList.forEach(handlerMock => handlerMock.mockClear());
-});
-
 test('will not let you construct it by itself', () => {
     expect(() => new BuildBus()).toThrowErrorMatchingSnapshot();
     expect(() => BuildBus.for('./fake-context')).not.toThrow();
 });
 
-test('caches buses for contexts', () => {
-    expect(BuildBus.for('./somewhere')).toBe(BuildBus.for('./somewhere'));
-    expect(mockHandlers.declares3.declare).toHaveBeenCalledTimes(1);
+test('caches buses for contexts and does not init the same bus twice', () => {
+    const bus = BuildBus.for('./somewhere');
+    const cachedBus = BuildBus.for('./somewhere');
+    bus.init();
+    cachedBus.init();
+    expect(bus).toBe(cachedBus);
+    expect(pertain).toHaveBeenCalledTimes(2);
+});
+
+test('can clear cache by context name', () => {
+    const bus = BuildBus.for('./somewhere');
+    expect(BuildBus.for('./somewhere')).toBe(bus);
+    BuildBus.clear('./somewhere');
+    expect(BuildBus.for('./somewhere')).not.toBe(bus);
 });
 
 test('calls declare and then intercept', () => {
```

Comment:
Would there be any value of including tests like these in the docs to describe how the underlying parts work? (Correct me if they already are :) )

Reply:
I've included examples of those tests into the new version of docs!